File: backend/app/api/v1/audit_visit_info.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.permissions import require_permission
from app.db.session import get_db
from app.models.user import User
from app.schemas.audit_visit_info import (
    AuditVisitInfoCreate,
    AuditVisitInfoListResponse,
    AuditVisitInfoMessageResponse,
    AuditVisitInfoResponse,
    AuditVisitInfoUpdate,
)
from app.services.audit_visit_info.audit_visit_info_service import (
    AuditVisitInfoService,
)


router = APIRouter(prefix="/audit-visit-info", tags=["Audit Visit Info"])


@router.get("", response_model=AuditVisitInfoListResponse)
async def list_audit_visit_info(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = None,
    is_active: bool | None = None,
    audit_id: int | None = None,
    team_id: int | None = None,
    status_filter: str | None = Query(default=None, alias="status"),
    sort_by: str = "visit_id",
    sort_order: str = "asc",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("menu.audit_visit_info.view")),
):
    service = AuditVisitInfoService(db)

    return await service.list_audit_visit_info(
        page=page,
        page_size=page_size,
        search=search,
        is_active=is_active,
        audit_id=audit_id,
        team_id=team_id,
        status_filter=status_filter,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.get("/{visit_id}", response_model=AuditVisitInfoResponse)
async def get_audit_visit_info(
    visit_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("menu.audit_visit_info.view")),
):
    service = AuditVisitInfoService(db)
    return await service.get_audit_visit_info(visit_id)


@router.post(
    "",
    response_model=AuditVisitInfoMessageResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_audit_visit_info(
    payload: AuditVisitInfoCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.audit_visit_info.create")),
):
    service = AuditVisitInfoService(db)

    return await service.create_audit_visit_info(
        payload=payload,
        created_by=current_user.user_id,
    )


@router.patch("/{visit_id}", response_model=AuditVisitInfoMessageResponse)
async def update_audit_visit_info(
    visit_id: int,
    payload: AuditVisitInfoUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.audit_visit_info.update")),
):
    service = AuditVisitInfoService(db)

    return await service.update_audit_visit_info(
        visit_id=visit_id,
        payload=payload,
        updated_by=current_user.user_id,
    )


@router.delete("/{visit_id}", response_model=AuditVisitInfoMessageResponse)
async def delete_audit_visit_info(
    visit_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.audit_visit_info.delete")),
):
    service = AuditVisitInfoService(db)

    return await service.deactivate_audit_visit_info(
        visit_id=visit_id,
        updated_by=current_user.user_id,
    )


@router.patch("/{visit_id}/restore", response_model=AuditVisitInfoMessageResponse)
async def restore_audit_visit_info(
    visit_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.audit_visit_info.restore")),
):
    service = AuditVisitInfoService(db)

    return await service.restore_audit_visit_info(
        visit_id=visit_id,
        updated_by=current_user.user_id,
    )


@router.delete(
    "/{visit_id}/permanent",
    response_model=AuditVisitInfoMessageResponse,
)
async def permanent_delete_audit_visit_info(
    visit_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(
        require_permission("api.audit_visit_info.permanent_delete")
    ),
):
    service = AuditVisitInfoService(db)
    return await service.permanent_delete_audit_visit_info(visit_id)

File: frontend/src/services/auditVisitInfo.ts
export type AuditVisitInfo = {
  visit_id: number;
  audit_id: number;
  team_id: number;
  client_address_id: number;
  visit_date: string;
  status: string;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditVisitInfoListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: AuditVisitInfo[];
};

export type AuditVisitInfoPayload = {
  audit_id: number;
  team_id: number;
  client_address_id: number;
  visit_date: string;
  status: string;
};

export type AuditVisitInfoUpdatePayload = Partial<AuditVisitInfoPayload> & {
  is_active?: boolean;
};

type ListParams = {
  page: number;
  pageSize: number;
  search?: string;
  isActive?: boolean;
  auditId?: number;
  teamId?: number;
  status?: string;
};

function buildQuery(params: ListParams) {
  const query = new URLSearchParams();

  query.set("page", String(params.page));
  query.set("page_size", String(params.pageSize));
  query.set("sort_by", "visit_id");
  query.set("sort_order", "desc");

  if (params.search) query.set("search", params.search);

  if (typeof params.isActive === "boolean") {
    query.set("is_active", String(params.isActive));
  }

  if (params.auditId) query.set("audit_id", String(params.auditId));
  if (params.teamId) query.set("team_id", String(params.teamId));
  if (params.status) query.set("status", params.status);

  return query.toString();
}

async function requestJson<T>(
  url: string,
  options?: RequestInit,
): Promise<T> {
  const response = await fetch(url, {
    ...options,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(options?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    const message =
      error?.detail || error?.message || "Audit Visit Info request failed.";

    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listAuditVisitInfo(params: ListParams) {
  const query = buildQuery(params);

  return requestJson<AuditVisitInfoListResponse>(
    `/api/backend/audit-visit-info?${query}`,
    {
      method: "GET",
    },
  );
}

export async function createAuditVisitInfo(payload: AuditVisitInfoPayload) {
  return requestJson<{ message: string; data: AuditVisitInfo }>(
    "/api/backend/audit-visit-info",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export async function updateAuditVisitInfo(
  id: number,
  payload: AuditVisitInfoUpdatePayload,
) {
  return requestJson<{ message: string; data: AuditVisitInfo }>(
    `/api/backend/audit-visit-info/${id}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export async function deactivateAuditVisitInfo(id: number) {
  return requestJson<{ message: string; data: AuditVisitInfo }>(
    `/api/backend/audit-visit-info/${id}`,
    {
      method: "DELETE",
    },
  );
}

export async function restoreAuditVisitInfo(id: number) {
  return requestJson<{ message: string; data: AuditVisitInfo }>(
    `/api/backend/audit-visit-info/${id}/restore`,
    {
      method: "PATCH",
    },
  );
}

export async function permanentDeleteAuditVisitInfo(id: number) {
  return requestJson<{ message: string; data: null }>(
    `/api/backend/audit-visit-info/${id}/permanent`,
    {
      method: "DELETE",
    },
  );
}
